database/mongo/controller.mjs

```diff
@@ -0,0 +1,237 @@
+            await this.client.connect()
+            console.log("MongoDB Connection Successful")
+            console.log(process.env.MONGODB)
+            return
+        } catch (err) {
+            console.error("MongoDB Connection Failed")
+            console.error(process.env.MONGODB)
+            console.error(err)
+            throw err
+        }
+    }
+
+    /** Close the connection with the mongo client */
+    async close() {
+        await this.client.close()
+        console.log("Mongo controller client has been closed")
+        return
+    }
+
+    /** 
+     * Generate an new mongo _id as a hex string (as opposed _id object, for example) 
+     * @return A hex string or error
+     * */
+    newID() {
+        try {
+            return new ObjectId().toHexString()
+        } catch (err) {
+            console.error(err)
+            return err
+        }
+    }
+
+    /** 
+     * Generally check that the controller has an active connection 
+     * @return boolean
+     * */
+    async connected() {
+        // Send a ping to confirm a successful connection
+        try {
+            let result = await this.db.command({ ping: 1 }).catch(err => { return false })
+            result = result.ok ? true : false
+            return result
+        } catch (err) {
+            console.error(err)
+            return false
+        }
+    }
+
+    /**
+     * Get by property matches and return all objects that match
+     * @param query JSON from an HTTP POST request.  It must contain at least one property.
+     * @return JSON Array of matched documents or standard error object
+     */
+    async read(query) {
+        try {
+            //need to determine what collection (projects, groups, userPerferences) this goes into.
+            const data_type = query["@type"] ?? query.type ?? null
+            if (!data_type)
+                return { "endpoint_error": "find", "status": 400, "message": `Cannot find 'type' on this data, and so cannot figure out a collection for it.` }
+            const collection = discernCollectionFromType(data_type)
+            if (!collection)
+                return { "endpoint_error": "find", "status": 400, "message": `Cannot figure which collection for object of type '${data_type}'` }
+            if (Object.keys(query).length === 0)
+                return { "endpoint_error": "find", "status": 400, "message": `Empty or null object detected.  You must provide a query object.` }
+            let result = await this.db.collection(collection).find(query).toArray()
+            return result
+        } catch (err) {
+            console.error(err)
+            return { "endpoint_error": "find", "status": 500, "message": "There was an error querying the database." }
+        }
+    }
+
+    /**
+     * Insert a document into the database (mongo)
+     * @param data JSON from an HTTP POST request
+     * @return The inserted document JSON or error JSON
+     */
+    async create(data) {
```

Comment:
Let's discuss this...

There are some patterns to consider. We want to return errors without breaking everything and we want to know what class of error it is. 
Inside this `try` we can throw an error and handle it all in a catch that returns

1. an Array like `[err, result]` that can be `[null, result]` for successes for an easy check.
2. an express-style `callback(err, result)` that is controlled by the caller.
3. an Object that always has `{ status, result }` that is easy to check.
4. some combination that does (3) if no callback is provided, for example.
5. just throw the Errors and let the caller handle it.